### ISSUE TITLE
file iteration

### ISSUE THREAD

**Issue by kylerlaird** [ninjaaron/replacing-bash-scripting-with-python#12]
I'm enjoying your document!  I've been moving away from Bash to Python scripting for years.  I'm always looking for better ways to do it.

I'm curious about the section on iterating through lines of a file.  You provide some relatively complicated ways to do it using "with", readline(), etc.  Why not simply "for line in open(file):"?

**Comment by ghost** [comment]
Nice work @kylerlaird that you are moving away from Bash.
I am copying from this [Python doc](https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files) and please read the details there.

"It is good practice to use the `with` keyword when dealing with file objects. The advantage is that the file is properly closed after its suite finishes, even if an exception is raised at some point. Using with is also much shorter than writing equivalent try-finally blocks..."

**Comment by kylerlaird** [comment]
The use of with is superfluous in this case.  The use of "my_file" is pointless.

These should be equivalent.
```
with open('my_file.txt') as my_file:
    for line in my_file:
        do_stuff_with(line.rstrip())
```

```
for line in open('my_file.txt'):
    do_stuff_with(line.rstrip())
```

**Comment by kylerlaird** [comment]
I agree that with is a good practice and a good tool to have (when needed for an object which is used multiple times).  The pattern of "loop through all of the lines in a file" is so common, though, that it's a shame to add extra crud to what should be simple in Python.

**Comment by ninjaaron** [comment]
It depends what you're doing. If you're simply looping over a single file, that's fine. However, if you're looping over multiple files, depending what happens with the garbage collector, you could run out of file descriptors from the OS. This would rarely happen in practice in CPython because it uses reference counting, but it's difficult to guarantee that it wouldn't happen in Pypy, which uses tracing GC.

Additionally, while the current implementation of `io` objects does clean up file descriptors when the object is destructed, it's not part of the documented semantics and shouldn't be relied upon.

You're right that `with` is frequently unnecessarily, and I don't always use it in my personal scripts. However, the whole point of this tutorial is help people write scripts that are _safe and maintainable_ by replacing Bash with Python. Running out of file descriptors on a corner case is exactly the kind of nonsense I'm trying to help people avoid. I'm not going to skip on a best-practice to avoid one extra line of code---especially if I'm teaching others.

**Comment by snahtajain** [comment]
You're correct that with is oftentimes superfluously, and I don't generally utilize it in my own contents. In any case, the general purpose of this instructional exercise is assist individuals with composing contents that are protected and viable by supplanting Bash with Python. Running out of record descriptors on a corner case is actually the sort of gibberish I'm attempting to help individuals evade. I'm not going to skirt on a best-practice to keep away from one additional line of code - particularly in case I'm instructing others.

I have some good books on python - 

**[Python Classes in Sangli](https://www.sevenmentor.com/python-course-in-sangli)**
**[Python Classes in Kolhapur](https://www.sevenmentor.com/python-course-in-kolhapur)**
**[Python Classes in Solapur](https://www.sevenmentor.com/python-course-in-solapur)**
**[Python Classes in Nagpur](https://www.sevenmentor.com/python-course-in-nagpur)**
**[Python Classes in latur](https://www.sevenmentor.com/python-course-in-latur)**
